### PR TITLE
build: infer platform from first node if none set

### PR DESCRIPTION
**Pull request by crazy-max**
closes https://github.com/docker/buildx/pull/2124

If user doesn't specify any platform, it should be inferred from the first node instead of using the local platform. In follow-up we should also look at the rules suggested in https://github.com/docker/buildx/issues/2119#issuecomment-1815517521 to take into account the preferred platform as well.

### CODE REVIEW

File: build/driver_test.go [hunk 1]
```diff
@@ -226,7 +226,7 @@ func TestSelectNodeCurrentPlatform(t *testing.T) {
 	require.NoError(t, err)
 	require.True(t, perfect)
 	require.Len(t, res, 1)
-	require.Equal(t, "bbb", res[0].Node().Builder)
+	require.Equal(t, "aaa", res[0].Node().Builder)
```
Comment (tonistiigi) [inline build/driver_test.go:229]: The real issue is that the platform in `opt` gets reset when it should not. We need to figure out a test to make sure it doesn't happen. Looks like the old build code only expect "resolveNode" to have a platform if it is an override to the existing platform but maybe this should be made safer (more explicit) somehow.

Comment (crazy-max) [reply build/driver_test.go:230]: Rebased on top of @jedevc changes in https://github.com/docker/buildx/compare/master...jedevc:buildx:fix-setting-wrong-platform-on-solve-opt to fix the solve opt case and remove the duplicated map.

Comment (crazy-max) [reply build/driver_test.go:230]: I guess you talk about the `TestSelectNodeNoPlatform` and default spec case right? It seems related to https://github.com/docker/buildx/issues/2119 and proposed new rules in https://github.com/docker/buildx/issues/2119#issuecomment-1815517521 should cover this in follow-up iiuc.

**Review by jedevc — APPROVED** [review]

File: build/driver.go [hunk 2]
```diff
@@ -19,6 +19,8 @@ import (
 )
 
 type resolvedNode struct {
+	SolveOpt *client.SolveOpt
```
Comment (tonistiigi) [inline build/driver.go:22]: I think @jedevc did not like the `driverPair` name, but this is the use case for why it was added. That we pass along a struct that contains both the driver (in here it is called resolver+platform, so driver for a specific node) and the request.

I think even if we use embed structs, the separation of structs would be clearer this way.

**Review by tonistiigi — COMMENTED** [review]
The logic side SGTM. Comment about the code organization.

File: build/driver.go (hunk 2 above)
Comment (crazy-max) [reply build/driver.go]: After discussing with @tonistiigi, the idea is to have a clear separation of concerns between drivers resolution and solve opts attached to the driver in build. Pushed an extra commit to move `*client.SolveOpt` to a local struct type in build.

**Review by tonistiigi — APPROVED** [review]

**Review by jedevc — APPROVED** [review]

**Pull request merged**